shared/src/commonMain/kotlin/dev/johnoreilly/confetti/utils/ClientQuery.kt

```diff
@@ -0,0 +1,42 @@
+package dev.johnoreilly.confetti.utils
+
+import com.apollographql.apollo3.ApolloCall
+import com.apollographql.apollo3.api.Error
+import com.apollographql.apollo3.api.Query
+import com.apollographql.apollo3.exception.ApolloException
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.runningFold
+
+object ClientQuery {
+    fun <D : Query.Data, T> ApolloCall<D>.toUiState(
+        onError: (List<Error>, ApolloException?) -> Unit = { _, _ -> },
+        initial: QueryResult<T> = QueryResult.Loading,
+        mapper: (D) -> T
+    ): Flow<QueryResult<T>> = toFlow().runningFold(initial) { previous, next ->
+        val apolloException = next.exception
+
+        if (next.hasErrors() || apolloException != null) {
+            onError(next.errors.orEmpty(), apolloException)
+        }
+
+        if (next.data != null) {
+            QueryResult.Success(mapper(next.data!!))
+        } else if (apolloException != null && previous is QueryResult.Loading) {
+            QueryResult.Error(apolloException)
```

Comment:
That doens't work with `CacheAndNetwork` and your first item is an error. It will return Error here while I think you want to ignore it?

Reply:
Yep, it's probably the type of thing that a very opinionated screen might want to decide.  But mostly here, I'm ok with

Loading -> Error -> Success

if it's uncommon. If success takes 3-4 seconds over the network, do you want it to show loading, or an error for a short while?

What do you suggest?